### PR TITLE
Support for tsirc alongside message tags. Close #149

### LINKED ISSUE #149: Support for tsirc alongside message tags.
Parser does not support instances where there are both messagetags and a TSIRC
timestamp.

Example:

```
«01:09:56» « In» BATCH backbuffer_#dmdirc_1486688995988 generic
«01:09:56» « In» :-BNC!bot@karatkievich.freenode.net NOTICE #dmdirc :Beginning backbuffer...
«01:09:56» « In» @batch=backbuffer_#dmdirc_1486688995988 @1486688970738@:Dataforce!dataforce@dataforce.org.uk PRIVMSG #dmdirc :test
«01:09:56» « In» :-BNC!bot@karatkievich.freenode.net NOTICE #dmdirc :End of backbuffer.
«01:09:56» « In» BATCH -backbuffer_#dmdirc_1486688995988
```

For testing purposes, both of these should be functionally identical:

```
@123@@tag=value :test ing
@tag=value @123@:test ing
```

### PR DESCRIPTION
**Pull request by ShaneMcC**
Eg, allow:
```
@123@@tag=value :test ing
@tag=value @123@:test ing
```

to both represent the same message.

### CODE REVIEW

File: irc/src/main/java/com/dmdirc/parser/irc/IRCReader.java [hunk 1]
```diff
@@ -251,30 +252,50 @@ public ReadLine(final String line, final String... lineTokens) {
...
+                }
+
+                tokens = new String[lineTokens.length - 1];
+                System.arraycopy(lineTokens, 1, tokens, 0, lineTokens.length - 1);
+            }
 
-                    tokens = new String[lineTokens.length - 1];
-                    System.arraycopy(lineTokens, 1, tokens, 0, lineTokens.length - 1);
+            // Look again for tsirc, as it may be after the message tags.
+            if (!tokens[0].isEmpty() && tokens[0].charAt(0) == '@') {
```
Comment (csmith) [inline irc/src/main/java/com/dmdirc/parser/irc/IRCReader.java]: Repeating this is icky :(

Can there be a tryParseTimestamp() method and a tryParseMessageTags() method we just call?

File: irc/src/main/java/com/dmdirc/parser/irc/IRCReader.java [hunk 2]
```diff
@@ -251,30 +252,50 @@ public ReadLine(final String line, final String... lineTokens) {
...
+                final String[] lineTags = tokens[0].substring(1).split(";");
+                for (final String keyVal : lineTags) {
+                    if (!keyVal.isEmpty()) {
+                        final String[] keyValue = keyVal.split("=", 2);
+                        tags.put(keyValue[0], keyValue.length > 1 ? keyValue[1] : "");
                     }
+                }
+
+                tokens = new String[lineTokens.length - 1];
```
Comment (csmith) [inline irc/src/main/java/com/dmdirc/parser/irc/IRCReader.java]: Why is the array shifting up here?

Seems like `@foo=bar@123@` will break. You missed that case from your tests, too :)

Comment (ShaneMcC) [inline irc/src/main/java/com/dmdirc/parser/irc/IRCReader.java]: `@foo=bar@123@` wouldn't be valid.

Either the whole first parameter is tags, and the TSIRC timestamp has been added to the start of the actual message or the TSRIC tag has been tacked on at the very beginning of the tags.

So:

```
@batch=backbuffer_#dmdirc_1486688995988 @1486688970738@:Dataforce!dataforce@dataforce.org.uk PRIVMSG #dmdirc :test
```
is what DFBnc does currently.

Or the alternative interpretation of the spec "at the very beginning of the line" would put it:

```
@1486688970738@@batch=backbuffer_#dmdirc_1486688995988 :Dataforce!dataforce@dataforce.org.uk PRIVMSG #dmdirc :test
```

There's no way to interpret it as putting the timestamp at the end of the list of tags.

Comment (ShaneMcC) [inline irc/src/main/java/com/dmdirc/parser/irc/IRCReader.java]: (Also, there's no real change to the code that parses the messages, just how it's called)

**Pull request merged**